File: Assets/Typescript~/src/Client/Controllers/Match/MatchController.ts
import { TabListController } from "@Easy/Core/Client/Controllers/TabList/TabListController";
import { FriendsController } from "@Easy/Core/Client/MainMenuControllers/Social/FriendsController";
import { Game } from "@Easy/Core/Shared/Game";
import { ColorUtil } from "@Easy/Core/Shared/Util/ColorUtil";
import { Signal } from "@Easy/Core/Shared/Util/Signal";
import { Theme } from "@Easy/Core/Shared/Util/Theme";
import { TimeUtil } from "@Easy/Core/Shared/Util/TimeUtil";
import { SetInterval } from "@Easy/Core/Shared/Util/Timer";
import { Controller, Dependency, OnStart } from "@easy-games/flamework-core";
import { ClientSignals } from "Client/ClientSignals";
import { MatchInfoDto } from "Shared/Match/MatchInfoDto";
import { MatchState } from "Shared/Match/MatchState";
import { Network } from "Shared/Network";

@Controller({})
export class MatchController implements OnStart {
	/** Initial state is always `MatchState.SETUP. */
	private state: MatchState = MatchState.SETUP;
	public matchStartTime: number | undefined;
	public matchInfo: MatchInfoDto | undefined;
	public eliminated = false;
	public onEliminated = new Signal<void>();

	constructor(private readonly tablistController: TabListController) {
		Network.ServerToClient.MatchInfo.client.OnServerEvent((matchInfoDto) => {
			this.matchInfo = matchInfoDto;
			this.state = matchInfoDto.matchState;
			if (matchInfoDto.matchStartTime !== undefined) {
				this.matchStartTime = matchInfoDto.matchStartTime;
			}
		});
		Network.ServerToClient.PlayerEliminated.client.OnServerEvent((clientId) => {
			if (clientId === Game.localPlayer.clientId) {
				this.eliminated = true;
				this.onEliminated.Fire();
			}
		});
	}

	OnStart(): void {
		/* Listen for match state change.  */
		Network.ServerToClient.MatchStateChange.client.OnServerEvent((newState, oldState) => {
			this.state = newState;
			/* Fire signal. */
			ClientSignals.MatchStateChange.Fire({ newState: this.state, oldState: oldState });
		});
		/* Listen for match start. */
		Network.ServerToClient.MatchStarted.client.OnServerEvent(() => {
			this.state = MatchState.RUNNING;
			this.matchStartTime = TimeUtil.GetServerTime();
			/* Fire signal */
			ClientSignals.MatchStart.Fire();
		});

		let timer = 0;
		if (this.matchStartTime !== undefined) {
			timer = math.floor(this.matchStartTime - TimeUtil.GetServerTime());
		}
		SetInterval(
			1,
			() => {
				if (this.state === MatchState.RUNNING) {
					timer++;
				}

				let minutes = math.floor(timer / 60);
				let seconds = math.floor(timer % 60);

				let time = string.format("%02d:%02d", minutes, seconds);

				let map = "";
				if (this.matchInfo) {
					map =
						ColorUtil.ColoredText(Theme.aqua, "<b>" + this.matchInfo.mapName + "</b>") +
						ColorUtil.ColoredText(Theme.gray, " by ") +
						ColorUtil.ColoredText(Theme.aqua, this.matchInfo.mapAuthors[0]);
				}

				let text =
					ColorUtil.ColoredText(Theme.white, `<b>BedWars.com</b>`) +
					ColorUtil.ColoredText(Theme.gray, "    Time: ") +
					ColorUtil.ColoredText(Theme.green, time) +
					"    " +
					map;
				this.tablistController.SetTitleText(text);
				if (this.matchInfo) {
					Dependency<FriendsController>().SetCustomGameTitle("BedWars | " + this.matchInfo?.mapName);
				} else {
					Dependency<FriendsController>().SetCustomGameTitle("BedWars | In Game");
				}
			},
			true,
		);
	}

	/**
	 * Returns current match state.
	 * @returns Current `MatchState`.
	 */
	public GetState(): MatchState {
		return this.state;
	}
}

File: Assets/Bundles/Client/Resources/TS/Controllers/Match/MatchController.lua
-- Compiled with unity-ts v2.1.0-75
local Reflect = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Reflect
local Controller = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Controller
local ClientSignals = require("Client/TS/ClientSignals").ClientSignals
local MatchState = require("Shared/TS/Match/MatchState").MatchState
local Network = require("Shared/TS/Network").Network
local MatchController
do
	MatchController = setmetatable({}, {
		__tostring = function()
			return "MatchController"
		end,
	})
	MatchController.__index = MatchController
	function MatchController.new(...)
		local self = setmetatable({}, MatchController)
		return self:constructor(...) or self
	end
	function MatchController:constructor()
		self.state = MatchState.SETUP
	end
	function MatchController:OnStart()
		-- Listen for match state change.
		Network.ServerToClient.MatchStateChange.Client:OnServerEvent(function(newState, oldState)
			self.state = newState
			-- Fire signal.
			ClientSignals.MatchStateChange:Fire({
				newState = self.state,
				oldState = oldState,
			})
		end)
		-- Listen for match start.
		Network.ServerToClient.MatchStarted.Client:OnServerEvent(function()
			self.state = MatchState.RUNNING
			-- Fire signal
			ClientSignals.MatchStart:Fire()
		end)
	end
	function MatchController:GetState()
		return self.state
	end
end
-- (Flamework) MatchController metadata
Reflect.defineMetadata(MatchController, "identifier", "Bundles/Client/Controllers/Match/MatchController@MatchController")
Reflect.defineMetadata(MatchController, "flamework:implements", { "$:flamework@OnStart" })
Reflect.decorate(MatchController, "$:flamework@Controller", Controller, { {} })
return {
	MatchController = MatchController,
}
-- ----------------------------------
-- ----------------------------------
